lib/Parse/ParseDecl.cpp

```diff
@@ -7719,9 +7719,12 @@ Parser::parseDeclOperator(ParseDeclOptions Flags, DeclAttributes &Attributes) {
   bool AllowTopLevel = Flags.contains(PD_AllowTopLevel);
 
   const auto maybeDiagnoseInvalidCharInOperatorName = [this](const Token &Tk) {
-    if (Tk.is(tok::identifier) &&
-        DeclAttribute::getAttrKindFromString(Tk.getText()) ==
-          DeclAttrKind::DAK_Count) {
+    if (Tk.is(tok::identifier) && Tk.getText().equals("$")) {
+      diagnose(Tk, diag::identifier_within_operator_name, Tk.getText());
+      return true;
+    } else if (Tk.is(tok::identifier) &&
+               DeclAttribute::getAttrKindFromString(Tk.getText()) ==
+                   DeclAttrKind::DAK_Count) {
```

Comment:
I think diagnosing both under a single `if (Tk.is(tok::identifier)) {` would be neater (the branches do not intersect either way).